src/sage/graphs/homomorphisms/count_homomorphisms.py

```diff
@@ -0,0 +1,236 @@
+        #   ((4, 0), (5, 2)): 20,
+        #   ((4, 1), (5, 0)): 30,
+        #   ((4, 1), (5, 2)): 40,
+        #   ((4, 2), (5, 0)): 50,
+        #   ((4, 2), (5, 1)): 60}, {}, ...]
+        self.DP_table = [{} for _ in range(len(self.dir_labelled_TD))]
+
+    def count_homomorphisms(self):
+        r"""
+        Return the number of homomorphisms from the graph `G` to the graph `H`.
+
+        A homomorphism from a graph `G` to a graph `H` is a function
+        `\varphi : V(G) \mapsto V(H)`, such that for any edge `uv \in E(G)` the
+        pair `\varphi(u) \varphi(v)` is an edge of `H`.
+
+        For more information, see the :wikipedia:`Graph_homomorphism`.
+
+        ALGORITHM:
+
+        This is an implementation based on the proof of Prop. 1.6 in [CDM2017]_.
+
+        OUTPUT:
+
+        - an integer, the number of homomorphisms from `graph` to `target_graph`
+
+        EXAMPLES::
+
+            sage: from sage.graphs.homomorphisms.count_homomorphisms import GraphHomomorphismCounter
+            sage: square = graphs.CycleGraph(4)
+            sage: bip = graphs.CompleteBipartiteGraph(2, 4)
+            sage: counter = GraphHomomorphismCounter(square, bip)
+            sage: counter.count_homomorphisms()
+            128
+        """
+        # Whether it's BFS or DFS, every node below join node(s) would be
+        # computed already, so we can go bottom-up safely.
+        for node in reversed(self.dir_labelled_TD.vertices()):
+            node_type = self.dir_labelled_TD.get_vertex(node)
+
+            match node_type:
+                case 'intro':
+                    self._add_intro_node(node)
+                case 'forget':
+                    self._add_forget_node(node)
+                case 'join':
+                    self._add_join_node(node)
+
+                case _:
+                    self._add_leaf_node(node)
+
+        return self.DP_table[0][0]
+
+    ### Main adding functions
+
+    def _add_leaf_node(self, node):
+        r"""
+        Add the computation result of processing leaf node(s) to the DP table.
+        """
+        node_index = get_node_index(node)
+        self.DP_table[node_index] = [1]
+
+    def _add_intro_node(self, node):
+        r"""
+        Add the computation result of processing introduce node(s) to the DP table.
+        """
+        # Basic setup
+        node_index, node_vertices = node
+        node_vtx_tuple = tuple(node_vertices)
+
+        child_node_index, child_node_vtx = self.dir_labelled_TD.neighbors_out(node)[0]
+        child_node_vtx_tuple = tuple(child_node_vtx)
+
+        mappings_length = self.actual_target_size ** len(node_vtx_tuple)
+
+        mappings_count = [0] * mappings_length
+
+        # Use the adjacency matrix when dense, otherwise the graph itself
+        target_density = self.actual_target_graph.density()
+        target = self.actual_target_graph.adjacency_matrix() if target_density >= self.density_threshold else self.actual_target_graph
```

Comment:
when using `adjacency_matrix`, it is safer to specify the order of the vertices using parameter `vertices` (see the documentation) to ensure that the i-th row/column of the matrix corresponds to the I-th vertex in your ordering.

Reply:
Do you mean something like in the current code?

Reply:
Please check the documentation of `adjacency_matrix`. You can specify the order of the vertices `adjacency_matrix(vertices=my_order)`,  where `my_order` is a list of vertices indicating some order.